### ISSUE TITLE
1:n sorting by different criteria (+ pagination)

### ISSUE THREAD

**Issue by langpavel** [join-monster/join-monster#29]
Hi there!

Thank you for :tada: awesome :100: project!

Are multiple sorting keys (and better, with pagination) supported?
It is possible to pass in argument which can be used to determine sorting key?

For example if I have **product** (1→n) **customer** relation and want sort all customers which buyed product by `date` *OR* `customer_name` (*OR* something different...) by [*Keyset Paging*](http://join-monster.readthedocs.io/en/latest/relay/#3-keyset-paging)

It is obvious that I need different cursors for each case.
From every cursor should be determined what pagination method was used in previous query *OR* I need sorting columns (and limit)
One example of pagination args: `first:5, sort: ["customer_name NULLS FIRST", "delivery_date DESC", "id"]`

Any clues or ideas how to implement this?

**Comment by acarl005** [comment]
Hi langpavel, thanks for the compliment!

Without pagination, dynamic sort order can be achieved by sorting in the `resolve` function. The team debated extensively whether we should support `ORDER BY` in the SQL. Our current stance is that Join Monster is for fetching the data. Once it has fetched the data, any additional presentation or sorting logic should be handled in the application. So just use JavaScript to sort. Of course, you can support dynamic sort orders based on the arguments.

Pagination will rely on a sort key to fetch the correct data. Currently, you are constrained to a single sort key. This constraint reflects a best practice from the database perspective. Most SQL databases require indexes for performant queries on a large number of paginated rows. Supporting multiple sort keys means multiple sets of database indexes, creating overhead. For this reason, choosing a single sort key is encouraged.

We don't want to be _too_ opinionated, though. Perhaps the `sortKey` could be a **function** that takes the field arguments..

``` js
someField: {
  sqlPaginate: true,
  sortKey: (args, context) => {
    return {
      order: args.orderBy,
      key: args.key
    }
  }
}
```

Would that satisfy your use case?

**Comment by langpavel** [comment]
Hi @acarl005, thanks for reply!

> Without pagination, dynamic sort order can be achieved by sorting in the resolve function.

Without pagination it is really not important :slightly_smiling_face:, it can be done on client side.

> Pagination will rely on a sort key to fetch the correct data. Currently, you are constrained to a single sort key. This constraint reflects a best practice from the database perspective...

I'm not a fool and I have indexes for every needed sorting criteria. Postgres is so powerful with way you can create indexes. But because you are not reflecting database schema in join-monster you don't know which indexes exists and can be used... So there should be some way to hint how to sort... Definitely should be possible to create composite sorting keys to deliver uniqueness for pagination. Most important for me is this (from Postgres documentation):

> An important special case is ORDER BY in combination with LIMIT n: an explicit sort will have to process all the data to identify the first n rows, but if there is an index matching the ORDER BY then the first n rows can be retrieved directly, without scanning the remainder at all.

**Comment by acarl005** [comment]
Multiple dynamic sort keys should be allowed. This feature has been added and published in v0.5.5.

Now, when doing offset pagination or keyset pagination, the `orderBy` or `sortKey` properties can be functions that return the order columns. The GraphQL args are the first parameter of this function. For example

``` js
    posts: {
      type: PostConnection, 
      args: connectionArgs,
      sqlPaginate: true,
      // sortKey could be an object... or a function that returns an object
      sortKey: graphqlArgs => {
        return {
          order: graphqlArgs.direction,
          key: graphqlArgs.column
        }
      },
      sqlJoin: (userTable, postTable) => `${userTable}.id = ${postTable}.author_id`
    },
```

You can have the GraphQL arguments determine the sort key. Simply declare the GraphQL arguments on the connection type and use the respective values in this function.
